akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/aa/AAAuctionExampleSpec.scala

```diff
@@ -0,0 +1,271 @@
+      initialBid: Bid, // the initial bid is basically the minimum price bidden at start time by the owner
+      closingAt: Instant,
+      responsibleForClosing: Boolean,
+      allDcs: Set[String])
+
+  def commandHandler(setup: AuctionSetup, ctx: ActorContext[AuctionCommand], aaContext: ActiveActiveContext)(
+      state: AuctionState,
+      command: AuctionCommand): Effect[AuctionEvent, AuctionState] = {
+    state.phase match {
+      case Closing(_) | Closed =>
+        command match {
+          case GetHighestBid(replyTo) =>
+            replyTo ! state.highestBid
+            Effect.none
+          case IsClosed(replyTo) =>
+            replyTo ! (state.phase == Closed)
+            Effect.none
+          case Finish =>
+            ctx.log.info("Finish")
+            Effect.persist(AuctionFinished(aaContext.replicaId))
+          case Close =>
+            ctx.log.info("Close")
+            require(shouldClose(setup, state))
+            // TODO send email (before or after persisting)
+            Effect.persist(WinnerDecided(aaContext.replicaId, state.highestBid, state.highestCounterOffer))
+          case _: OfferBid =>
+            // auction finished, no more bids accepted
+            Effect.unhandled
+        }
+      case Running =>
+        command match {
+          case OfferBid(bidder, offer) =>
+            Effect.persist(
+              BidRegistered(
+                Bid(bidder, offer, Instant.ofEpochMilli(aaContext.currentTimeMillis()), aaContext.replicaId)))
+          case GetHighestBid(replyTo) =>
+            replyTo ! state.highestBid
+            Effect.none
+          case Finish =>
+            Effect.persist(AuctionFinished(aaContext.replicaId))
+          case Close =>
+            ctx.log.warn("Premature close")
+            // Close should only be triggered when we have already finished
+            Effect.unhandled
+          case IsClosed(replyTo) =>
+            replyTo ! false
+            Effect.none
+        }
+    }
+  }
+
+  private def shouldClose(auctionSetup: AuctionSetup, state: AuctionState): Boolean = {
+    auctionSetup.responsibleForClosing && (state.phase match {
+      case Closing(alreadyFinishedAtDc) =>
+        val allDone = auctionSetup.allDcs.diff(alreadyFinishedAtDc).isEmpty
+        if (!allDone) {
+          println(
+            s"Not closing auction as not all DCs have reported finished. All DCs: ${auctionSetup.allDcs}. Reported finished ${alreadyFinishedAtDc}")
+        }
+        allDone
+      case _ =>
+        false
+    })
+  }
+
+  def eventHandler(ctx: ActorContext[AuctionCommand], aaCtx: ActiveActiveContext, setup: AuctionSetup)(
+      state: AuctionState,
+      event: AuctionEvent): AuctionState = {
+
+    val newState = state.applyEvent(event)
+    ctx.log.infoN("Applying event {}. New start {}", event, newState)
+    if (!aaCtx.recoveryRunning) {
+      eventTriggers(setup, ctx, aaCtx, event, newState)
+    }
+    newState
+
+  }
+
+  private def eventTriggers(
```

Comment:
this is the main API difference from multi dc persistence. There isn't an event trigger that can return an effect, instead you must send a command back to self if you want to persist other events

Reply:
I think this approach is fine. An alternative would be to register an eventTrigger function `Event => Command` in the ActiveActiveContext. It would still be processed as a Command, but would not be used during replay. Not a big advantage, and could be added later if we see people struggle with it.